web/modules/custom/joinup_communities/tallinn/src/Plugin/Block/DashboardBlock.php

```diff
@@ -0,0 +1,319 @@
+  protected $entityTypeManager;
+
+  /**
+   * The dashboard access service.
+   *
+   * @var \Drupal\tallinn\DashboardAccessInterface
+   */
+  protected $dashboardAccess;
+
+  /**
+   * The config factory service.
+   *
+   * @var \Drupal\Core\Config\ConfigFactoryInterface
+   */
+  protected $configFactory;
+
+  /**
+   * Constructs a new plugin class instance.
+   *
+   * @param array $configuration
+   *   A configuration array containing information about the plugin instance.
+   * @param string $plugin_id
+   *   The plugin_id for the plugin instance.
+   * @param mixed $plugin_definition
+   *   The plugin implementation definition.
+   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entity_type_manager
+   *   The entity type manager service.
+   * @param \Drupal\tallinn\DashboardAccessInterface $dashboard_access
+   *   The dashboard access service.
+   * @param \Drupal\Core\Config\ConfigFactoryInterface $config_factory
+   *   The config factory service.
+   */
+  public function __construct(array $configuration, $plugin_id, $plugin_definition, EntityTypeManagerInterface $entity_type_manager, DashboardAccessInterface $dashboard_access, ConfigFactoryInterface $config_factory) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+    $this->entityTypeManager = $entity_type_manager;
+    $this->dashboardAccess = $dashboard_access;
+    $this->configFactory = $config_factory;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $container->get('entity_type.manager'),
+      $container->get('tallinn.dashbord.access'),
+      $container->get('config.factory')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function build(): array {
+    $build = $this->buildRenderArray();
+    $this->addSelectors($build);
+    return [$build];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getCacheTags(): array {
+    // Add the collection cache tags.
+    $tags = Cache::mergeTags(parent::getCacheTags(), Rdf::load(TALLINN_COMMUNITY_ID)->getCacheTags());
+
+    // Merge the tags of each report.
+    $storage = $this->entityTypeManager->getStorage('node');
+    foreach ($storage->loadByProperties(['type' => 'tallinn_report']) as $report) {
+      $tags = Cache::mergeTags($tags, $report->getCacheTags());
+    }
+
+    // Add the cache tags of 'tallinn.settings' object.
+    return Cache::mergeTags($tags, $this->configFactory->get('tallinn.settings')->getCacheTags());
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function blockAccess(AccountInterface $account): AccessResultInterface {
+    return $this->dashboardAccess->access($account);
+  }
+
+  /**
+   * Builds the block render array.
+   *
+   * @return array
+   *   Render array
+   */
+  protected function buildRenderArray(): array {
```

Comment:
There is a lot of very intricate building of HTML in this block, but I can't see it when I am running the tests. There is also no test that checks the content of this block as far as I can see. Is this out of scope for the ticket?

Reply:
This HTML markup is coming from the INNO team, see https://codepen.io/weslito/pen/ajObKq?editors=0010. I only abstracted it as a Drupal render array.